tools/copyblocks/main.go

```diff
@@ -0,0 +1,507 @@
+		return errors.Wrapf(err, "failed to list tenants")
+	}
+
+	return concurrency.ForEachUser(ctx, tenants, cfg.tenantConcurrency, func(ctx context.Context, tenantID string) error {
+		if !isAllowedUser(enabledUsers, disabledUsers, tenantID) {
+			return nil
+		}
+
+		logger := log.With(logger, "tenantID", tenantID)
+
+		blocks, err := listBlocksForTenant(ctx, sourceBucket, tenantID)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to list blocks for tenant", "err", err)
+			return errors.Wrapf(err, "failed to list blocks for tenant %v", tenantID)
+		}
+
+		markers, err := listBlockMarkersForTenant(ctx, sourceBucket, tenantID, cfg.destBucket)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to list blocks markers for tenant", "err", err)
+			return errors.Wrapf(err, "failed to list block markers for tenant %v", tenantID)
+		}
+
+		var blockIDs []string
+		for _, b := range blocks {
+			blockIDs = append(blockIDs, b.String())
+		}
+
+		// We use ForEachUser here to keep processing other blocks, if the block fails. We pass block IDs as "users".
+		return concurrency.ForEachUser(ctx, blockIDs, cfg.blocksConcurrency, func(ctx context.Context, blockIDStr string) error {
+			blockID, err := ulid.Parse(blockIDStr)
+			if err != nil {
+				return err
+			}
+
+			logger := log.With(logger, "block", blockID)
+
+			// Skip if the block was already copied.
+			if markers[blockID].copied {
+				level.Debug(logger).Log("msg", "skipping block because it has been copied already")
+				return nil
+			}
+
+			blockMeta, err := loadMetaJSONFile(ctx, sourceBucket, tenantID, blockID)
+			if err != nil {
+				level.Error(logger).Log("msg", "skipping block, failed to read meta.json file", "err", err)
+				return err
+			}
+
+			// Add min/max time to each log entry. This is useful for debugging purposes.
+			blockMinTime := time.Unix(0, blockMeta.MinTime*int64(time.Millisecond)).UTC()
+			blockMaxTime := time.Unix(0, blockMeta.MaxTime*int64(time.Millisecond)).UTC()
+			logger = log.With(logger, "block_min_time", blockMinTime, "block_max_time", blockMaxTime)
+
+			if markers[blockID].deletion {
+				level.Debug(logger).Log("msg", "skipping block because it is marked for deletion")
+				return nil
+			}
+
+			// If the min time filter is set, only blocks with MinTime >= the configured value are copied.
+			if filterMinTime := time.Time(cfg.minTime); !filterMinTime.IsZero() && blockMinTime.Before(filterMinTime) {
+				level.Debug(logger).Log("msg", "skipping block, block min time is lower than the configured min time filter", "configured_min_time", filterMinTime)
+				return nil
+			}
+
+			// If the max time filter is set, only blocks with MaxTime <= the configured value are copied.
+			if filterMaxTime := time.Time(cfg.maxTime); !filterMaxTime.IsZero() && blockMaxTime.After(filterMaxTime) {
+				level.Debug(logger).Log("msg", "skipping block, block max time is greater than the configured max time filter", "configured_max_time", filterMaxTime)
```

Comment:
Nit here, but there is "lower" and "greater" in these strings. For consistency, use either "lower" and "higher" or "less than" and "greater than".

Reply:
Addressed, thanks.